server/configuration.go

```diff
@@ -1,31 +1,70 @@
 package main
 
-import "errors"
+import (
+	"github.com/pkg/errors"
+)
 
+// Config captures the plugin's external configuration as exposed in the Mattermost server
+// configuration, as well as values computed from the configuration. Any public fields will be
+// deserialized from the Mattermost server configuration in OnConfigurationChange.
 type Config struct {
 	Trigger string
 }
 
+// OnConfigurationChange loads the plugin configuration, validates it and saves it.
 func (p *MatterpollPlugin) OnConfigurationChange() error {
-	c := &Config{}
-	if err := p.API.LoadPluginConfiguration(c); err != nil {
-		return err
+	config := new(Config)
+
+	if err := p.API.LoadPluginConfiguration(config); err != nil {
+		return errors.Wrap(err, "failed to load plugin configuration")
 	}
 
-	if c.Trigger == "" {
+	if config.Trigger == "" {
 		return errors.New("Empty trigger not allowed")
 	}
 
 	if p.Config != nil {
 		if err := p.API.UnregisterCommand("", p.Config.Trigger); err != nil {
-			return err
+			return errors.Wrap(err, "failed to unregister old command")
 		}
 	}
-	if err := p.API.RegisterCommand(getCommand(c.Trigger)); err != nil {
-		return err
+	if err := p.API.RegisterCommand(getCommand(config.Trigger)); err != nil {
+		return errors.Wrap(err, "failed to register new command")
 	}
 
 	p.ServerConfig = p.API.GetConfig()
-	p.Config = c
+	p.setConfiguration(config)
 	return nil
 }
+
+// getConfiguration retrieves the active configuration under lock, making it safe to use
+// concurrently. The active configuration may change underneath the client of this method, but
+// the struct returned by this API call is considered immutable.
+func (p *MatterpollPlugin) getConfiguration() *Config {
+	p.configurationLock.RLock()
+	defer p.configurationLock.RUnlock()
```

Comment:
Sorry, I don't know too much about mutex, so I don't understand what does this lock improve.

If calling RUnlock here, RLock is immediately unlocked in the end of this method.
So I think that we can update config value at any time regardless of this changes.

```
config := p.getConfiguration()
config.Trigger = "dummy"
```

Is this PR intended to prevent config values from being changed during evaluating simple if statement and return statement?

Reply:
> If calling RUnlock here, RLock is immediately unlocked in the end of this method.
So I think that we can update config value at any time regardless of this changes.

That is is correct. A client of this method can change the configuration value at any time. But besides of a separate package, the is no way to avoid this. The comment of this method explains this quite well
> getConfiguration retrieves the active configuration under lock, making it safe to use
concurrently. The active configuration may change underneath the client of this method, but
the struct returned by this API call is considered immutable.

> Is this PR intended to prevent config values from being changed during evaluating simple if statement and return statement?

Correct. This PR ensured that a configuration doesn't get changed while a methods evaluates it. The biggest problem is in `ExecuteCommand()`. We evaluate the configuration values multiple times, but a change can happen every moment.

Reply:
It all makes sense now.
Thank you for the clarification!